src/main/scala/mesosphere/marathon/state/MarathonStore.scala

```diff
@@ -1,139 +1,75 @@
-            Some(stateFromBytes(newVar.value))
-          case None =>
-            throw new StorageException(s"Failed to store $key")
-        }
+  def modify(key: String)(f: (() => S) => S): Future[S] = withLock(key) {
+    val res = store.load(prefix + key).flatMap {
+      case Some(entity) =>
+        bytesRead.update(entity.bytes.length)
+        val updated = f(() => stateFromBytes(entity.bytes))
+        val updatedEntity = entity.mutate(updated.toProtoByteArray)
+        bytesWritten.update(updatedEntity.bytes.length)
+        store.save(updatedEntity)
       case None =>
-        throw new StorageException(s"Failed to read $key, does not exist, should have been created automatically.")
+        val created = f(() => newState()).toProtoByteArray
+        bytesWritten.update(created.length)
+        store.create(prefix + key, created)
     }
-
-    res onComplete { _ =>
-      lock.release()
-    }
-
-    res
+    res.map { entity => stateFromBytes(entity.bytes) }
   }
 
-  def expunge(key: String): Future[Boolean] = {
-    val lock = locks.get(key)
-    lock.acquire()
-
-    val res = state.fetch(prefix + key) flatMap {
-      case Some(variable) =>
-        bytesRead.update(Option(variable.value).map(_.length).getOrElse(0))
-        state.expunge(variable) map {
-          case Some(b) => b.booleanValue()
-          case None    => throw new StorageException(s"Failed to expunge $key")
-        }
-
-      case None => throw new StorageException(s"Failed to read $key")
-    }
-
-    res onComplete { _ =>
-      lock.release()
-    }
-
-    res
+  def expunge(key: String): Future[Boolean] = withLock(key) {
+    store.delete(prefix + key) map { _ => true }
   }
 
-  def names(): Future[Iterator[String]] = {
-    BackToTheFuture.futureToFuture(state.names()).map {
-      _.asScala.collect {
-        case name if name startsWith prefix =>
-          name.replaceFirst(prefix, "")
+  def names(): Future[Seq[String]] = {
+    store.allIds().map {
+      _.collect {
+        case name if name startsWith prefix => name.replaceFirst(prefix, "")
       }
-    }.recover {
-      case NonFatal(e) =>
-        // TODO: Currently this code path is taken when the zookeeper path does not exist yet. It would be nice
-        // to not log this as a warning.
-        //
-        // Unfortunately, this results in a NullPointerException in `throw e.getCause()` in BackToTheFuture because
-        // the native mesos code returns an ExecutionException without cause. Therefore, we cannot robustly
-        // differentiate between exceptions which are "normal" and exceptions which indicate real errors
-        // and we have to log them all.
-        log.warn(s"exception while calling $getClass.names() for prefix $prefix", e)
-        Iterator.empty
     }
   }
 
   private def stateFromBytes(bytes: Array[Byte]): S = {
     newState().mergeFromProto(bytes)
   }
+
+  private def withLock[T](key: String)(future: => Future[T]): Future[T] = {
```

Comment:
Really nice that you have extracted that.

Optimally, LockManager would return a LockManger object containing this method, though.